subsys/bluetooth/services/ras/rrsp/ras_rd_buffer.c

```diff
@@ -0,0 +1,293 @@
+{
+	if (buf->conn) {
+		bt_conn_unref(buf->conn);
+	}
+
+	buf->conn = NULL;
+	buf->ready = false;
+	buf->busy = false;
+	buf->acked = false;
+	buf->refcount = 0;
+	buf->subevent_cursor = 0;
+	atomic_clear(&buf->refcount);
+}
+
+static struct ras_rd_buffer *rd_buffer_alloc(struct bt_conn *conn, uint16_t ranging_counter)
+{
+	uint16_t conn_buffer_count = 0;
+	uint16_t oldest_ranging_counter = UINT16_MAX;
+	struct ras_rd_buffer *available_free_buffer = NULL;
+	struct ras_rd_buffer *available_oldest_buffer = NULL;
+
+	for (uint8_t i = 0; i < RD_POOL_SIZE; i++) {
+		if (rd_buffer_pool[i].conn == conn) {
+			conn_buffer_count++;
+
+			/* Only overwrite buffers that have ranging data stored
+			 * and are not being read.
+			 */
+			if (rd_buffer_pool[i].ready && !rd_buffer_pool[i].busy &&
+			    atomic_get(&rd_buffer_pool[i].refcount) == 0 &&
+			    rd_buffer_pool[i].ranging_counter < oldest_ranging_counter) {
+				oldest_ranging_counter = rd_buffer_pool[i].ranging_counter;
+				available_oldest_buffer = &rd_buffer_pool[i];
+			}
+		}
+
+		if (available_free_buffer == NULL && rd_buffer_pool[i].conn == NULL) {
+			available_free_buffer = &rd_buffer_pool[i];
+		}
+	}
+
+	/* Allocate the buffer straight away if the connection has not reached
+	 * the maximum number of buffers allocated.
+	 */
+	if (conn_buffer_count < CONFIG_BT_RAS_RRSP_RD_BUFFERS_PER_CONN) {
+		rd_buffer_init(conn, available_free_buffer, ranging_counter);
+
+		return available_free_buffer;
+	}
+
+	/* Overwrite the oldest stored ranging buffer that is not in use */
+	if (available_oldest_buffer != NULL) {
+		if (!available_oldest_buffer->acked) {
+			/* Only notify if the peer has not read the buffer yet. */
+			notify_rd_overwritten(conn, oldest_ranging_counter);
+		}
+		rd_buffer_free(available_oldest_buffer);
+
+		rd_buffer_init(conn, available_oldest_buffer, ranging_counter);
+
+		return available_oldest_buffer;
+	}
+
+	/* Could not allocate a buffer */
+	return NULL;
+}
+
+static void cs_procedure_enabled(struct bt_conn *conn,
+				 struct bt_conn_le_cs_procedure_enable_complete *params)
+{
+	uint8_t conn_index = bt_conn_index(conn);
+
+	__ASSERT_NO_MSG(conn_index < ARRAY_SIZE(tx_power_cache));
+
+	tx_power_cache[conn_index] = params->selected_tx_power;
+}
+
+static bool process_step_data(struct bt_le_cs_subevent_step *step, void *user_data)
+{
+	struct ras_rd_buffer *buf = (struct ras_rd_buffer *)user_data;
+
+	buf->procedure.subevents[buf->subevent_cursor] = step->mode;
+	buf->subevent_cursor += BT_RAS_STEP_MODE_LEN;
+
+	memcpy(&buf->procedure.subevents[buf->subevent_cursor], step->data, step->data_len);
+	buf->subevent_cursor += step->data_len;
+
+	return true;
+}
+
+static void subevent_data_available(struct bt_conn *conn,
+				    struct bt_conn_le_cs_subevent_result *result)
+{
+	LOG_DBG("Procedure %u", result->header.procedure_counter);
+	struct ras_rd_buffer *buf =
+		rd_buffer_get(conn, result->header.procedure_counter, false, true);
+
+	if (!buf) {
```

Comment:
Revisit this - there's a race condition here when the app returns the last available buffer while CS subevents are being received.